### ISSUE TITLE
Single object detection and boxing

### ISSUE THREAD

**Issue by carlogrisetti** [ludwig-ai/ludwig#331]
I'm having a use case where I do not only have to classify an image on a (small) set of categories, but I also need to do a really basic object detection, that can be accomplished via the image classification itself, since it will always be a single object in the image, but I will also need to have a rough estimate on where a specific feature of this object is, in order to perform further analysis on that part of the image.
I cannot complete this using other techniques, because the object I'm focusing on can be easily identified by a human, but has subtle variances between each instance that forces the solution into an AI-driven approach.

Is there some simplification in implementing some sort of single object detection into Ludwig, as opposite to a multi-object detection, or can that be naively trained with an image input and adding some numeric outputs (for ROI coordinates) ?

In the latter, a simple custom "box" output type extension, that wraps around existing numerical output features could be useful in abstracting this to a higher level.

I know you are focused on bugs, and I could take on this extension, but I wanted to pick your mind in the feasibility of such an approach.

Bye!

**Comment by w4nderlust** [comment]
A box output type (or more generic vector potentially) would be a nice addition.
One could train it with the sum of mse on each dimension and use iou or jaccard index as a measure of quality.
It shouldn't be hard to implement, if you are interested we can chat about it.
Moreover, I believe using [CoordConv](https://arxiv.org/abs/1807.03247) would be greatly beneficial for the quality of the results. It will be super simple to implement it.

As a REALLY SUBOPTIMAL hacky solution, you can have 4 numerical output features for x, y, h and w respectively and see how far can you get with it.

**Comment by carlogrisetti** [comment]
Really suboptimal but i'm having great results with a simple numerical output, with initializer: ones
I really hope that's not only due of my net having memorized the inputs, because images are so similar, so I'm waiting for a bigger dataset to be ready, and in the meantime I'm playing with hyperparameters.

I took a look at what you proposed, but it's really out of my league to implement :)

Pull the "help wanted" tag, and just leave this as enhancement if you want

**Comment by nitrofortin** [comment]
I would be interested in implementing this enhancement. Could you just point me where the solution would live? Thanks!

**Comment by w4nderlust** [comment]
@nitrofortin  thanks for offering to help!
The ideal place to put it would be in the `features` package, something like `boundingbox_feature.py`.
It would have to contain both a class improting from `BaseFeature` and one importing from `OutputFeature` and should implement all the methods that are defined in those objects.
You can use the `numerical_feature.py` as a blueprint and modify things accordingly, knowing that the outputs are going to be a 4d vector, that the measures are different (intersection-over-union) and the loss will be different.
Refer to the object detection literature (for instance eq. 2 and 3 in [Fast R-CNN](https://www.cv-foundation.org/openaccess/content_iccv_2015/papers/Girshick_Fast_R-CNN_ICCV_2015_paper.pdf) or section 2.1 in [YOLOv3](https://arxiv.org/pdf/1804.02767.pdf)) for figuring out the measures people usually use and the most adequate losses.
Ideally one may want to support also multiple boxes rather than only one, but this would be a great starting point.
Let me know if you need further guidance.

**Comment by nitrofortin** [comment]
Thanks for the details, will be working on it!

**Comment by nitrofortin** [comment]
Submitted a PR. Now it needs to pass the tests, work with an object detection problem (instead of the dummy iris dataset used) and supplemented with other losses (e.g.: like YOLOv3 as you mentioned).

**Comment by w4nderlust** [comment]
@nitrofortin thank you so much! Will take a look over the weekend.
